### PR TITLE
support both spring-boot 2.5 and spring-boot 2.6 in pipeline

### PR DESCRIPTION
**Pull request by Netyyyy**
1. Add script of change spring-boot version

### CODE REVIEW

File: sdk/spring/ci.yml [hunk 1]
```diff
@@ -1,337 +1,537 @@
...
-      - sdk/spring/spring-cloud-azure-starter-stream-eventhubs/pom.xml
-      - sdk/spring/spring-cloud-azure-starter-stream-servicebus/pom.xml
-      - sdk/spring/spring-cloud-azure-starter/pom.xml
-      - sdk/spring/spring-cloud-azure-stream-binder-eventhubs-core/pom.xml
-      - sdk/spring/spring-cloud-azure-stream-binder-eventhubs/pom.xml
-      - sdk/spring/spring-cloud-azure-stream-binder-servicebus-core/pom.xml
-      - sdk/spring/spring-cloud-azure-stream-binder-servicebus/pom.xml
-      - sdk/spring/spring-cloud-azure-stream-binder-test/pom.xml
-      - sdk/spring/spring-cloud-azure-test-servicebus-binder/pom.xml
+    branches:
```
Comment (chenrujun) [inline sdk/spring/ci.yml:4]: Could you please avoid unnecessary change.
Here, use 2 spaces as indent.

File: sdk/spring/scripts/ci_add_spring_boot_dependencies_management.py [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+#    `python .\sdk\spring\scripts\ci_add_spring_boot_dependencies_management.py`.
+#
+# The script must be run at the root of azure-sdk-for-java.
+
+import os
+import time
+
+
+SPRING_BOOT_DEPENDENCIES_VERSION = '2.6.1'
+SPRING_CLOUD_DEPENDENCIES_VERSION = '2021.0.0'
```
Comment (chenrujun) [inline sdk/spring/scripts/ci_add_spring_boot_dependencies_management.py:18]: Make the 2 variable to parameter.

File: sdk/spring/scripts/ci_add_spring_boot_dependencies_management.py [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+
+def add_dependency_management(c1,c2):
```
Comment (chenrujun) [inline sdk/spring/scripts/ci_add_spring_boot_dependencies_management.py:20]: Make variable name readable. `c1, c2` is not a good variable name.

File: sdk/spring/scripts/ci_add_spring_boot_dependencies_management.py [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+    cores += '\n        <version>'+SPRING_BOOT_DEPENDENCIES_VERSION+'</version><!-- version test for ci -->'
+    cores += '\n        <type>pom</type>'
+    cores += '\n        <scope>import</scope>'
+    cores += '\n      </dependency>\n'
+    cores += '\n      <dependency>'
+    cores += '\n        <groupId>org.springframework.cloud</groupId>'
+    cores += '\n        <artifactId>spring-cloud-dependencies</artifactId>'
+    cores += '\n        <version>'+SPRING_CLOUD_DEPENDENCIES_VERSION+'</version><!-- version test for ci -->'
+    cores += '\n        <type>pom</type>'
+    cores += '\n        <scope>import</scope>'
```
Comment (chenrujun) [inline sdk/spring/scripts/ci_add_spring_boot_dependencies_management.py:55]: Python supports string in multiline.
Refs: https://www.techbeamers.com/python-multiline-string/

File: sdk/spring/scripts/ci_add_spring_boot_dependencies_management.py [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+    cores = '\n        <groupId>org.springframework.boot</groupId>'
+    cores += '\n        <artifactId>spring-boot-dependencies</artifactId>'
+    cores += '\n        <version>'+SPRING_BOOT_DEPENDENCIES_VERSION+'</version><!-- version test for ci -->'
+    cores += '\n        <type>pom</type>'
+    cores += '\n        <scope>import</scope>'
+    cores += '\n      </dependency>\n'
+    cores += '\n      <dependency>'
+    cores += '\n        <groupId>org.springframework.cloud</groupId>'
+    cores += '\n        <artifactId>spring-cloud-dependencies</artifactId>'
+    cores += '\n        <version>'+SPRING_CLOUD_DEPENDENCIES_VERSION+'</version><!-- version test for ci -->'
```
Comment (chenrujun) [inline sdk/spring/scripts/ci_add_spring_boot_dependencies_management.py:53]: Use string format instead of concatenate string.

Example:
```
'org.springframework.boot:spring-boot-starter-parent;{}'.format(SPRING_BOOT_VERSION),
```

File: sdk/spring/scripts/ci_change_version.py [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+# 6. Change 'LAST_SPRING_BOOT_VERSION' to the last version in this script manually.
+# 7. Then the ci will automatically run command `python .\sdk\spring\scripts\ci_change_version.py`.
+#
+# The script must be run at the root of azure-sdk-for-java.
+
+import os
+import time
+
+SPRING_BOOT_DEPENDENCIES_VERSION = '2.6.1'
+LAST_SPRING_BOOT_DEPENDENCIES_VERSION = '2.5.4'
```
Comment (chenrujun) [inline sdk/spring/scripts/ci_change_version.py:23]: As discussed, delete this file, and use parameter instead.

File: sdk/spring/scripts/ci_update_versions.py [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+    else:
+        for root, _, files in os.walk("./sdk/spring"):
+            for file_name in files:
+                file_path = root + os.sep + file_name
+                if  file_name.startswith('pom') and file_name.endswith('.xml'):
+                    update_versions(update_type, version_map, ext_dep_map, file_path)
+
+def main():
+    parser = argparse.ArgumentParser(description='Replace version numbers in poms.')
+    parser.add_argument('--update-type', '--ut', type=UpdateType, choices=list(UpdateType))
```
Comment (chenrujun) [inline sdk/spring/scripts/ci_update_versions.py:180]: this parameter is not necessary, we only update external dependencies.

File: sdk/spring/scripts/ci_update_versions.py [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+    parser.add_argument('--target-file', '--tf', nargs='?', help='File to update (optional) - all files in the current directory and subdirectories are scanned if omitted')
```
Comment (chenrujun) [inline sdk/spring/scripts/ci_update_versions.py:181]: > All files in the current directory

Not `current directory.`


We do not need to update specific file. How about do like this:

```
    parser.add_argument(
        '--targer-folder',
        type=str,
        required=False,
        default='./sdk/string',
        help='Set target folder.'
    )
```

File: sdk/spring/scripts/ci_update_versions.py [hunk 9]
```diff
@@ -0,0 +1,190 @@
...
+
+SPRING_BOOT_MANAGED_EXTERNAL_DEPENDENCIES_FILE_NAME = 'sdk/spring/scripts/spring_boot_{}_managed_external_dependencies.txt'.format(SPRING_BOOT_VERSION)
+
+def update_versions(update_type, version_map, ext_dep_map, target_file):
+    #list for new pom
+    newlines = []
+    repl_open, repl_line, file_changed, is_include = False, False, False, False
+    print('processing: ' + target_file)
+    try:
+        with open(target_file, encoding='utf-8') as f:
```
Comment (chenrujun) [inline sdk/spring/scripts/ci_update_versions.py:40]: How about use replace instead of create new file?
Just like what we did in `sync_external_dependencies.py`.

```
    with in_place.InPlace(target_file) as file:
        for line in file:
            line = line.strip()
            if line.startswith('#') or not line:
                file.write(line)
```

File: sdk/spring/scripts/ci_update_versions.py [hunk 10]
```diff
@@ -0,0 +1,190 @@
...
+                                raise ValueError('Module: {0} does not have an external dependency version.\nFile={1}\nLine={2}'.format(module_name, target_file, line))
+                    else:
+                        raise ValueError('Invalid version type: {} for module: {}.\nFile={}\nLine={}'.format(version_type, module_name, target_file, line))
+
+                    newlines.append(newline)
+                    if line != newline:
+                        file_changed = True
+                else:
+                    newlines.append(line)
+
```
Comment (chenrujun) [inline sdk/spring/scripts/ci_update_versions.py:116]: Delete unnecessary empty line.

File: sdk/spring/scripts/ci_update_versions.py [hunk 11]
```diff
@@ -0,0 +1,190 @@
...
+        if file_changed:
+            with open(target_file, 'w', encoding='utf-8') as f:
+                for line in newlines:
+                    f.write(line)
+
+    except Exception as e:
+        print("Unexpected exception: " + str(e))
+        traceback.print_exc(file=sys.stderr)
+
+def load_version_map_from_file(the_file, version_map):
```
Comment (chenrujun) [inline sdk/spring/scripts/ci_update_versions.py:129]: Could you please refer to `sync_external_dependencies.py`:
```
    with open(source_file) as file:
        for line in file:
            line = line.strip()
            if line.startswith('#') or not line:
                file.write(line)
            else:
                key_value = line.split(';', 1)
                key = key_value[0]
                value = key_value[1]
                dependency_dict[key] = value
```

File: sdk/spring/scripts/ci_update_versions.py [hunk 12]
```diff
@@ -0,0 +1,190 @@
...
+        print('external_dependency_file=' + dependency_file)
+        load_version_map_from_file(dependency_file, ext_dep_map)
+
+    display_version_info(version_map)
+    display_version_info(ext_dep_map)
+
+    if target_file:
+        update_versions(update_type, version_map, ext_dep_map, target_file)
+    else:
+        for root, _, files in os.walk("./sdk/spring"):
```
Comment (chenrujun) [inline sdk/spring/scripts/ci_update_versions.py:172]: It should be `target_file` instead of `"./sdk/spring"`.

File: sdk/spring/scripts/ci_update_versions.py [hunk 13]
```diff
@@ -0,0 +1,190 @@
...
+
+            version_map[module.name] = module
+
+def display_version_info(version_map):
+    for value in version_map.values():
+        print(value)
+
+def update_versions_all(update_type, target_file):
+    # dictionary
+    version_map = {}
```
Comment (chenrujun) [inline sdk/spring/scripts/ci_update_versions.py:158]: `version_map` is always empty?

File: sdk/spring/scripts/ci_update_versions.py [hunk 14]
```diff
@@ -0,0 +1,190 @@
...
+    ext_dep_map = {}
```
Comment (chenrujun) [inline sdk/spring/scripts/ci_update_versions.py:159]: Avoid use confusing variable name. How about renaming to `external_dependency_version_map`?

File: sdk/spring/scripts/ci_update_versions.py [hunk 15]
```diff
@@ -0,0 +1,190 @@
...
+                    tempName = module.name[len('unreleased_'):]
+                # if there isn't a non beta or unreleased entry then raise an issue
+                if tempName not in version_map:
+                    raise ValueError('Version file: {0} does not contain a non-beta or non-unreleased entry for beta_/unreleased_ library: {1}'.format(the_file, module.name))
+
+            version_map[module.name] = module
+
+def display_version_info(version_map):
+    for value in version_map.values():
+        print(value)
```
Comment (chenrujun) [inline sdk/spring/scripts/ci_update_versions.py:154]: Could you please use log instead of `print`?
Refer to `get_spring_boot_managed_external_dependencies.py`:

```
    parser.add_argument(
        '--log',
        type=str,
        choices=['debug', 'info', 'warn', 'error', 'none'],
        required=False,
        default='info',
        help='Set log level.'
    )
    args = parser.parse_args()
    log_dict = {
        'debug': Log.DEBUG,
        'info': Log.INFO,
        'warn': Log.WARN,
        'error': Log.ERROR,
        'none': Log.NONE
    }
    log.set_log_level(log_dict[args.log])
    # log.log_level_test()
```

File: sdk/spring/scripts/ci_utils.py [hunk 16]
```diff
@@ -0,0 +1,102 @@
+# some classes, methods support for ci_update_versions.py
```
Comment (chenrujun) [inline sdk/spring/scripts/ci_utils.py:1]: How about move contents into `ci_update_versions.py`?

File: sdk/spring/scripts/ci_utils.py [hunk 17]
```diff
@@ -0,0 +1,102 @@
...
+
+from enum import Enum
+import re
+
+include_update_marker = re.compile(r'\{x-include-update;([^;]+);([^}]+)\}')
+version_update_start_marker = re.compile(r'\{x-version-update-start;([^;]+);([^}]+)\}')
+version_update_end_marker = re.compile(r'\{x-version-update-end\}')
```
Comment (chenrujun) [inline sdk/spring/scripts/ci_utils.py:8]: The 2 is never used.

File: sdk/spring/scripts/ci_update_versions.py [hunk 18]
```diff
@@ -0,0 +1,190 @@
...
+                    module_name, version_type = match.group(1), match.group(2)
+                    repl_line = True
+                #return the first match{x-include-update;([^;]+);([^}]+)\}
+                elif include_update_marker.search(line):
+                    match = include_update_marker.search(line)
+                    module_name, version_type = match.group(1), match.group(2)
+                    repl_line = True
+                    is_include = True
+                else:
+                    #return the first match{x-version-update-start;([^;]+);([^}]+)\}
```
Comment (chenrujun) [inline sdk/spring/scripts/ci_update_versions.py:56]: This is used in README.md, we do not need this logic.

File: sdk/spring/scripts/ci_utils.py [hunk 19]
```diff
@@ -0,0 +1,102 @@
...
+
+# External dependency versions do not have to match semver format and the semver regular expressions
+# will partially match and produce some hilarious results.
+external_dependency_version_regex = r'(?<=<version>).+?(?=</version>)'
+
+
+# Allow list prefix remover
+allowlist_exception_identifier_remover_regex = re.compile(r'^(?:.+_)(?=.+:)(.*)$')
+
+class UpdateType(Enum):
```
Comment (chenrujun) [inline sdk/spring/scripts/ci_utils.py:30]: Delete this, because we only need update external_dependency.

File: sdk/spring/scripts/ci_utils.py [hunk 20]
```diff
@@ -0,0 +1,102 @@
...
+        return self.value
+
+    # defining string is necessary to get ArgumentParser's help output to produce
+    # human readable values of BuildType
+    def __str__(self):
+        return self.value
+
+class CodeModule:
+    def __init__(self, module_str):
+        # For library versions there will be up to 3 items resulting from the split
```
Comment (chenrujun) [inline sdk/spring/scripts/ci_utils.py:47]: We do not need this, because we only need to update external dependencies' version.

**Review by chenrujun — CHANGES_REQUESTED** [review]

File: sdk/spring/scripts/ci_update_versions.py [hunk 21]
```diff
@@ -0,0 +1,104 @@
...
+                match = version_update_marker.search(line)
+                module_name, version_type = match.group(1), match.group(2)
+                if module_name in external_dependency_version_map and version_type == 'external_dependency':
+                    new_version = external_dependency_version_map[module_name]
+                    newline = re.sub(external_dependency_version_regex, new_version, line)
+                    file.write(newline)
+                    log.info("updating " + target_file + ":" + newline)
+                else:
+                    file.write(line)
+            elif include_update_marker.search(line):
```
Comment (chenrujun) [inline sdk/spring/scripts/ci_update_versions.py:42]: It's possible that `version_update_marker.search(line) == True && include_update_marker.search(line) == True.`, so it should not be `elif`.

Comment (Netyyyy) [reply sdk/spring/scripts/ci_update_versions.py:42]: They are two different conditions that need different processing methods,   ``version_update_marker.search(line)`` will be replace with ``external_dependency_version_map[module_name]`` but ``include_update_marker.search(line)`` will be replace with ``module_name + ':[' + external_dependency_version_map[module_name] + ']'`` , so I think they shouldn't be written together or connected with '&&'.

File: sdk/spring/scripts/ci_update_versions.py [hunk 22]
```diff
@@ -0,0 +1,104 @@
...
+                match = include_update_marker.search(line)
```
Comment (chenrujun) [inline sdk/spring/scripts/ci_update_versions.py:43]: Can we use log like this :
```
for marker in [version_update_marker, include_update_marker]:
  ...
```
to avoid duplicate code.

File: sdk/spring/ci-template.yml [hunk 23]
```diff
@@ -0,0 +1,31 @@
+jobs:
```
Comment (saragluna) [inline sdk/spring/ci-template.yml:1]: Are we using 4 spaces indent here? How about changing to 2?

**Review by saragluna — CHANGES_REQUESTED** [review]

File: sdk/spring/spring-sdk-client.yml [hunk 24]
```diff
@@ -0,0 +1,131 @@
...
+#          SourceDirectory: $(Build.SourcesDirectory)
+#          BasePathLength: 38
+
+#      - ${{if ne(parameters.SDKType, 'data')}}:
+#          - ${{ each artifact in parameters.Artifacts }}:
+#              - ${{if ne(artifact.skipVerifyChangelog, 'true')}}:
+#                  - template: /eng/common/pipelines/templates/steps/verify-changelog.yml
+#                    parameters:
+#                      PackageName: ${{artifact.name}}
+#                      ServiceName: ${{parameters.ServiceDirectory}}
```
Comment (saragluna) [inline sdk/spring/spring-sdk-client.yml:92]: delete them?

File: sdk/spring/ci.yml [hunk 25]
```diff
@@ -55,7 +55,6 @@ trigger:
       - sdk/spring/spring-cloud-azure-stream-binder-eventhubs/pom.xml
       - sdk/spring/spring-cloud-azure-stream-binder-servicebus-core/pom.xml
       - sdk/spring/spring-cloud-azure-stream-binder-servicebus/pom.xml
-      - sdk/spring/spring-cloud-azure-test-eventhubs-binder/pom.xml
```
Comment (saragluna) [inline sdk/spring/ci.yml:58]: The spring-cloud-azure-test-eventhubs-binder is added back to the 4.0 branch. https://github.com/Azure/azure-sdk-for-java/tree/feature/azure-spring-cloud-4.0/sdk/spring

File: sdk/spring/ci.yml [hunk 26]
```diff
@@ -113,12 +112,15 @@ pr:
...
```
Comment (saragluna) [inline sdk/spring/ci.yml:116]: same

File: sdk/spring/spring-sdk-client.yml [hunk 27]
```diff
@@ -0,0 +1,296 @@
...
+    type: string
+    default: 2.5.7
+  - name: LastSpringCloudVersion
+    type: string
+    default: 2020.0.4
+
+stages:
+  - stage: Build
+    jobs:
+      - template: /eng/pipelines/templates/jobs/ci.yml
```
Comment (saragluna) [inline sdk/spring/spring-sdk-client.yml:66]: can we use ../../eng/pipelines/templates/stages/archetype-sdk-client.yml here?

File: sdk/spring/spring-sdk-client.yml [hunk 28]
```diff
@@ -0,0 +1,296 @@
...
+          MatrixConfigs:
+            - ${{ each config in parameters.MatrixConfigs }}:
+                - ${{ config }}
+            - ${{ each config in parameters.AdditionalMatrixConfigs }}:
+                - ${{ config }}
+          MatrixFilters:
+            - ${{ each filter in parameters.MatrixFilters }}:
+                - ${{ filter }}
+            # Skip TestFromSource jobs for SDKType data
+            - ${{ if eq(parameters.SDKType, 'data') }}:
```
Comment (saragluna) [inline sdk/spring/spring-sdk-client.yml:206]: There are some duplicates in these two stages, can we extract the common template for them?

**Review by saragluna — CHANGES_REQUESTED** [review]

**Review by saragluna — APPROVED** [review]
LGTM.

File: sdk/spring/spring-sdk-client-test.yml [hunk 29]
```diff
@@ -0,0 +1,140 @@
...
+    default: []
+  - name: SpringBootVersion
+    type: string
+    default: 'not-specified' # Set a default that breaks in obvious ways.
+  - name: SpringCloudVersion
+    type: string
+    default: 'not-specified' # Set a default that breaks in obvious ways.
+
+jobs:
+  - template: /eng/common/pipelines/templates/jobs/archetype-sdk-tests-generate.yml
```
Comment (chenrujun) [inline sdk/spring/spring-sdk-client-test.yml:48]: How about not using this template? Just do these steps: 
1. update version
2. inset dependency management
3. run test

**Review by chenrujun — APPROVED** [review]
LGTM.
